### DISCUSSION TITLE
How do I control cell padding in typst documents?

### DISCUSSION THREAD

**Discussion by kathrine-mcaulay** [quarto-dev/quarto-cli#9090] [Q&A] [answered]
### Description

I was writing this as a bug report, but let me first make sure that I'm not just being silly:

I am able to set table inset for a table where I explicitly outline the contents (first example below), but I cannot seem to get this to pass to a table derived from a saved dataframe (second example). I included the fill call as a sanity check that I'm using `set table` appropriately.

Note that I am passing this through `gt::gt` here, but I have also tried `kableExtra::kable`. I had initially started with the `{tinytable}` package, but the reason I moved away from this was because I didn't seem to have a way to control cell padding there either.

Is this a bug? And/or is there another way to control this? 

````qmd
---
title: "Untitled"
format: typst
execute: 
  echo: false
---

```{=typst}
#set table(
    inset: 20pt,
    fill: luma(225)
  )
  
#table(
  [Hello],
  [World],
)
```

```{r reprex table}
#| label: tbl-reprex
#| tbl-cap: Example
#| tbl-cap-location: bottom 

gt::gt(head(mtcars)[,1:2])

```
````

- IDE: RStudio 2023.12.1 Build 402 RStudio 2023.12.1+402 "Ocean Storm"
- OS: Windows 10 22H2 build 19045-4046

**Comment by mcanouil** [comment]
Note that `gt`, `kable` (`kableExtra` only knows about LaTeX and HTML) does not support Typst, so they emit markdown tables.
Markdown tables are mostly (if not entirely) handled by Pandoc.

To debug Typst, you can use `keep-typ: true` in your YAML front matter and refers to [Typst documentation](https://typst.app/docs/) to see what in your document you need to fix.

PS: note that in the below snippet from your code you are defining the label twice. In Quarto, you have no reason to have anything in the code cell header, all options exist in YAML which is the recommended way of passing options.

````qmd
```{r reprex table}
#| label: tbl-reprex
#| tbl-cap: Example
#| tbl-cap-location: bottom 

gt::gt(head(mtcars)[,1:2])
```
````

To

````qmd
```{r}
#| label: tbl-reprex
#| tbl-cap: Example
#| tbl-cap-location: bottom 

gt::gt(head(mtcars)[,1:2])
```
````

I am not aware of any R packages currently supporting Typst directly.

Because of Pandoc and R packages (or any third party actually) lack of Typst support, I ended up writing the table myself in an extension: https://github.com/mcanouil/quarto-invoice/blob/main/template.qmd
You can see the grid markdown table which allow merging cells, etc. but unfortunately not supported by Pandoc for Typst format.

**Comment by kathrine-mcaulay** [comment]
Thanks!

I guess my question now is: why is my set inset being overridden with this 6pt inset? Where is that coming from?

If this _is_ something that comes from the `{gt}` formatting that I cannot control and, then I guess I need to take my question over to the `{tinytable}` pages to determine why I can't seem to use their `bootstrap_css_rule()` to edit my `tt` tables in Typst either.

```
#set table(
    inset: 20pt,
    fill: luma(225)
  )
  
#table(
  [Hello],
  [World],
)
#block[
#figure([
#block[
#block[
#figure(
align(center)[#table(
  columns: 2,
  align: (col, row) => (auto,auto,).at(col),
  inset: 6pt,
  [mpg], [cyl],
  [21.0],
  [6],
  [21.0],
  [6],
  [22.8],
  [4],
  [21.4],
  [6],
  [18.7],
  [8],
  [18.1],
  [6],
)]
)

]
]
], caption: figure.caption(
position: bottom, 
[
Example
]), 
kind: "quarto-float-tbl", 
supplement: "Table", 
numbering: "1", 
)
<tbl-reprex>


]
```

And as an aside, thanks for the note on labels. I had been using YAML label for table and figure cross-referencing, and the code block header to name my chunks. Today I learned that the YAML label does indeed also name the chunk, which inherently makes sense, but I ever thought about it, so thank you!

**Reply by mcanouil** [reply, marked as answer]
CSS is HTML, this has absolutely no effect (and never will) in Typst or LaTeX.
As I said and as far as I know, there is no R packages supporting Typst, so in the end they either emit markdown or HTML tables which are processed by Quarto to AST (Pandoc representation of a document) then transformed to the appropriate format.
You can use `keep-md: true` to see the intermediate markdown document that is used.
Additionally, you can use `html-table-processing: none` (you might need 1.5 for this) in your YAML front matter to disable Quarto processing of HTML table to see if it is related to what you see.

See https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/typst/pandoc/quarto/template.typ for the template and partials.
See https://quarto.org/docs/output-formats/typst.html and https://quarto.org/docs/output-formats/typst-custom.html

**Reply by kathrine-mcaulay** [reply]
This is incredibly helpful. 

`html-table-processing: none` just eliminates the table from the output.

I tried adding `as_raw_html` from `{gt}` thinking that perhaps I may then have another avenue to modify the html in that interim stage, but  this did not seem to change the behavior.

Marking this one as the answer since keeping the .typ and .md files definitely helped me see what was wrong, even if I was unable to correct it! So thanks for your help with that!

In the end, my work-around is to write my data to csv and then load it back via typst using `let`. This is clunky, but it then lets me use some of the more recently added native table styling features. Specifically, [this guide](https://typst.app/docs/guides/table-guide/) was just added, which has some nice examples.

````
---
title: "Untitled"
format: typst
execute: 
  echo: false
keep-typ: true
keep-md: true
---

```{=typst}
#set table(
    inset: 20pt,
    fill: luma(225)
  )
  
#table(
  [Hello],
  [World],
)


```

```{r}

gt::gt(head(mtcars)[,1:2]) |>
  write.csv("cars.csv")

```

```{=typst}

#let cars = csv("cars.csv")

#table(
  columns: 2,
  ..for (.., mpg, cyl) in cars {
    (mpg, cyl)
  }
)

```
````

**Reply by cscheid** [reply]
@kathrine-mcaulay Thanks a really slick trick. I will just note that the new table styling features for typst literally landed [today](https://typst.app/docs/changelog/), so we haven't incorporated it yet :)

**Reply by kathrine-mcaulay** [reply]
Noted! 😅

I'm relieved enough that I can finally read the contents of my tables without squinting at this point. I'll just look forward to styling them at a later date!

**Reply by mcanouil** [reply]
You don't need `gt` if you are going to export as CSV.
The following is enough.

```r
write.csv(x = head(mtcars)[,1:2], file = "cars.csv")
```

Additionally, to output markdown table `knitr::kable()` is what you want.

Regarding table definition, the issue is really on Pandoc side as it converts the table.
See the document below, will exhibit the issue.

````qmd
---
format: typst
include-in-header:
  text: |
    #set table(
      inset: 40pt,
      fill: luma(225)
    )
---

## Bad

| Hello |
|-------|
| World |


## Good

```{=typst}
#table(
  [Hello],
  [World],
)
```
````

For instance a slight modification of the above

````qmd
---
format: typst
---

```{=typst}
#set table(
  inset: 40pt,
  fill: luma(225)
)
```

## Bad

| Hello |
|-------|
| World |


## Good

```{=typst}
#table(
  [Hello],
  [World],
)
```
````

Using Pandoc will show you that it is the source of the inset addition:

```sh
quarto pandoc index.qmd --from markdown --to typst
```

```typst
#set table(
  inset: 40pt,
  fill: luma(225)
)
== Bad
<bad>
#figure(
align(center)[#table(
  columns: 1,
  align: (col, row) => (auto,).at(col),
  inset: 6pt,
  [Hello],
  [World],
)]
)

== Good
<good>
#table(
  [Hello],
  [World],
)
```

FYI I opened an issue in Pandoc to remove the hardcoded `inset` which I don't think should be.

- https://github.com/jgm/pandoc/issues/9580

**Reply by kathrine-mcaulay** [reply]
> You don't need `gt` if you are going to export as CSV. The following is enough.
> 
> ```r
> write.csv(x = head(mtcars)[,1:2], file = "cars.csv")
> ```

Literally just noticed this too.

Thanks for clarifying and opening the issue.

**Reply by mcanouil** [reply]
FYI, the hardcoded inset value has been removed in Pandoc.
It will be available in a future version of Pandoc then later on Quarto.
